### ISSUE TITLE
Question about the reproduced result

### ISSUE THREAD

**Issue by tdd233** [TOPO-EPFL/CrossLoc#2]
Hi,

Thanks for the code, I'm very interested in your outstanding work!!!
I have tried to retrain your network, and the result shows that the index values of in-place localization accuracy I reproduced over urbanscape dataset are better than those given in your paper. So I want to ask you what might cause this situation.

Looking forward to your reply.
********************************************************************************************************************************************
_Here is the result over urbanscape dataset from your paper:_
![image](https://user-images.githubusercontent.com/93235196/230622072-d49dd192-7ad3-4b9a-a83a-50d361e30c4a.png)
_Here is quantitative comparison between paper's result and my reproduced result:_
![image](https://user-images.githubusercontent.com/93235196/230624448-9c1bb98a-0623-4580-a78d-9198675f1f2f.png)

**Comment by qiyan98** [comment]
Hi, 

Thanks for your message and reproduction! I think there are three reasons that may lead to this discrepancy:
* general training stochasticities, like different pytorch/cuda versions or OS or different GPU architecture. Our papers' results were produced using a single NVIDIA V100 32GB GPU without distributed training.
* checkpoint selection rule. We were using the performance on the validation set to select the best checkpoint during training. I believe a _better_ model could be found if we directly evaluated its performance on the testing set.
* moreover, the PnP solver may include another level of randomness. We borrow the improved PnP solver from the [dsacstar](https://github.com/TOPO-EPFL/CrossLoc/tree/main/dsacstar) as it is. Previously, we also found that the PnP solver could give different results on different machines (with the same predicted scene coordinates).

Lastly, we provided our model checkpoints at https://1drv.ms/u/s!AnkbqTET-eNqgoRsgBXkEg-PFSqudA?e=S6Pf43. You can also give it a try to rule out some possibilities.
